### PR TITLE
refactor(engine): replace collect_errors with finalize

### PR DESCRIPTION
**Pull request by rgrinberg**
The behavior should be equivalent but [finalize] is a little cleaner and
faster.

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

### CODE REVIEW

**Review by Alizter — APPROVED** [review]

**Comment by snowleopard** [comment]
> The behavior should be equivalent but [finalize] is a little cleaner and faster.

I agree that it's cleaner but how is it faster? `Fiber.finalize` calls `collect_errors` on `finally`, whereas the original code doesn't.

What do benchmarks say?

**Comment by snowleopard** [comment]
I also doubt the "should be equivalent" part. Surely the error-collection behaviour is changing?

**Comment by rgrinberg** [comment]
> I agree that it's cleaner but how is it faster? Fiber.finalize calls collect_errors on finally, whereas the original code doesn't.

I thought `finalize` was a primitive in fiber. Indeed it's not and there's no performance difference.

> What do benchmarks say?

They say there's no difference. Not surprisingly.

> I also doubt the "should be equivalent" part. Surely the error-collection behaviour is changing?

Yes, and actually I think this new behavior is improved. In the code before this PR, an IO error deleting the sandbox or deleting the pending targets would cause the original errors from running the actions to be discarded.

**Comment by snowleopard** [comment]
> > I also doubt the "should be equivalent" part. Surely the error-collection behaviour is changing?
> 
> Yes, and actually I think this new behavior is improved. In the code before this PR, an IO error deleting the sandbox or deleting the pending targets would cause the original errors from running the actions to be discarded.

Yep, that sounds right. Maybe it's worth adding a test that for this case? (Feel free to skip if it's difficult to write such a test.)

**Review by snowleopard — APPROVED** [review]

**Comment by rgrinberg** [comment]
It ended up being a little difficult to write a test. The only scenario I could think of is the user accidentally meddling with the sandbox.

**Comment by snowleopard** [comment]
> The only scenario I could think of is the user accidentally meddling with the sandbox.

...which is not really uncommon and providing good error messages in this case seems useful.

**Comment by snowleopard** [comment]
Imported this PR internally.

@rgrinberg What's the meaning of the `upstream-sync` label? Shall I remove it now that I've imported the PR?

File: src/dune_engine/build_system.ml [hunk 1]
```diff
@@ -380,53 +380,47 @@ end = struct
...
-            let should_be_skipped =
-              match rule_kind with
-              | Normal_rule -> fun (_ : Path.Build.t) -> false
-              | Anonymous_action { stamp_file; _ } -> Path.Build.equal stamp_file
+    Fiber.finalize
+      ~finally:(fun () ->
+        match sandbox with
+        | Some sandbox -> Sandbox.destroy sandbox
+        | None ->
+          (* All went well, these targets are no longer pending. *)
```
Comment (snowleopard) [inline src/dune_engine/build_system.ml:388]: Btw, the "all went well" part is a little confusing (the action might have failed actually). Internally, I've just deleted this comment as it seems unnecessary.

**Comment by snowleopard** [comment]
> Imported this PR internally.

Actually, importing this PR turned out pretty annoying because internally we have the code below. Note that we run `Js_corrected_files.rename_corrected_files` only in the error case. This makes it necessary to change `Fiber.finalize` to communicate the execution result to `finally`, which is a somewhat invasive change.

```
    let () =
      match exec_result, sandbox with
      | Error _, Some sandbox -> Js_corrected_files.rename_corrected_files sandbox
      | _ -> ()
    in
    let* () =
      match sandbox with
      | Some sandbox -> Sandbox.destroy sandbox
      | None ->
        (* All went well, these targets are no longer pending. *)
        Pending_targets.remove targets;
        Fiber.return ()
    in
    match exec_result with
    | Ok res -> Fiber.return res
    | Error l -> Fiber.reraise_all l
```

**Comment by snowleopard** [comment]
OK, never mind the last point. I figured out where to move the `Js_corrected_files.rename_corrected_files` bit without introducing a more complicated version of `Fiber.finalize`.

**Comment by rgrinberg** [comment]
> What's the meaning of the upstream-sync label? Shall I remove it now that I've imported the PR?

Yup. That's the goal. Just a list we can use to manage the PR's that are still pending.

File: src/dune_engine/build_system.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_engine/build_system.ml:388]: Would you like to submit a PR to do the same here?

**Comment by snowleopard** [comment]
> Yup. That's the goal. Just a list we can use to manage the PR's that are still pending.

Good idea. Thanks!

File: src/dune_engine/build_system.ml (hunk 1 above)
Comment (snowleopard) [reply src/dune_engine/build_system.ml:388]: Done: https://github.com/ocaml/dune/pull/8808

**Pull request merged**